api/authenticator.py

```diff
@@ -335,23 +335,48 @@ def set_authorization_identifier(self, authorization_identifier):
         self.authorization_identifiers = authorization_identifiers
 
 class Authenticator(object):
+    """Route requests to the appropriate LibraryAuthenticator.
+    """
+
+    def __init__(self, _db):
+        self.library_authenticators = {}
+
+        for library in _db.query(Library):
+            self.library_authenticators[library.short_name] = LibraryAuthenticator.from_config(_db, library)
+
+    def authenticated_patron(self, _db, header):
+        short_name = flask.request.library.short_name
+        if short_name not in self.library_authenticators:
+            return LIBRARY_NOT_FOUND
+        return self.library_authenticators[short_name].authenticated_patron(_db, header)
+
+    def create_authentication_document(self):
+        short_name = flask.request.library.short_name
+        if short_name not in self.library_authenticators:
+            return LIBRARY_NOT_FOUND
+        return self.library_authenticators[short_name].create_authentication_document()
+
+    def create_authentication_headers(self):
+        short_name = flask.request.library.short_name
+        if short_name not in self.library_authenticators:
+            return LIBRARY_NOT_FOUND
+        return self.library_authenticators[short_name].create_authentication_headers()
+
+    def get_credential_from_header(self, header):
+        short_name = flask.request.library.short_name
+        if short_name not in self.library_authenticators:
+            return LIBRARY_NOT_FOUND
+        return self.library_authenticators[short_name].get_credential_from_header(header)
+
```

Comment:
I don't know if it's worth the effort, but you could remove a lot of this code with a helper method like this:

```
def invoke_authenticator_method(self, method_name, *args, **kwargs):
    short_name = flask.request.library.short_name
    if short_name not in self.library_authenticators:
        return LIBRARY_NOT_FOUND
    return getattr(self.library_authenticators[short_name], method_name)(*args, **kwargs)
```